drivers/soundwire/intel.c

```diff
@@ -1412,6 +1412,108 @@ static int intel_master_remove(struct sdw_master_device *md)
 
 #ifdef CONFIG_PM
 
+static int _suspend(struct device *dev, bool clock_stop)
+{
+	struct sdw_cdns *cdns = dev_get_drvdata(dev);
+	struct sdw_intel *sdw = cdns_to_intel(cdns);
+	int ret;
+
+	if (cdns->bus.prop.hw_disabled) {
+		dev_dbg(dev, "SoundWire master %d is disabled, ignoring\n",
+			cdns->bus.link_id);
+		return 0;
+	}
+
+	dev_err(dev, "%s start\n", __func__);
+
+	if (clock_stop) {
+		ret = sdw_cdns_suspend(cdns, true);
+		if (ret < 0) {
+			dev_err(dev, "cannot enable clock stop on suspend\n");
+			return ret;
+		}
+	}
+
+	ret = sdw_cdns_enable_interrupt(cdns, false);
+	if (ret < 0) {
+		dev_err(dev, "cannot disable interrupts on suspend\n");
+		return ret;
+	}
+
+	ret = intel_link_power_down(sdw);
+	if (ret) {
+		dev_err(dev, "Link power down failed: %d", ret);
+		return ret;
+	}
+
+	/* set WAKEEN interrupt for wake-up events */
+	intel_shim_wake(sdw, clock_stop);
+
+	dev_err(dev, "%s done\n", __func__);
+
+	return 0;
+}
+
+static int _resume(struct device *dev, bool clock_stop)
+{
+	struct sdw_cdns *cdns = dev_get_drvdata(dev);
+	struct sdw_intel *sdw = cdns_to_intel(cdns);
+	int ret;
+
+	if (cdns->bus.prop.hw_disabled) {
+		dev_dbg(dev, "SoundWire master %d is disabled, ignoring\n",
+			cdns->bus.link_id);
+		return 0;
+	}
+
+	dev_err(dev, "%s start\n", __func__);
+
+	/* Invoke shim for wake disable */
+	intel_shim_wake(sdw, false);
+
+	ret = intel_init(sdw);
+	if (ret) {
+		dev_err(dev, "%s failed: %d", __func__, ret);
+		return ret;
+	}
+
+	/*
+	 * make sure all Slaves are tagged as UNATTACHED and provide
+	 * reason for reinitialization
+	 */
+	sdw_clear_slave_status(&sdw->cdns.bus,
+			       SDW_UNATTACH_REQUEST_MASTER_RESET);
```

Comment:
no, this should only be done when you don't use clock-stop mode.

Reply:
please check the following explanation for sdw_cdns_exit_reset. Without this function, I got IO timeout

Reply:
NO. you cannot use this 
````c
	/* use hardware generated reset */
	cdns_updatel(cdns, CDNS_MCP_CONTROL,
		     CDNS_MCP_CONTROL_HW_RST,
		     CDNS_MCP_CONTROL_HW_RST);
````

it has to become dependent on the clock stop mode

Reply:
thanks, I will refine it.

Reply:
I will try the merging irq patch, maybe it will improve my issue.